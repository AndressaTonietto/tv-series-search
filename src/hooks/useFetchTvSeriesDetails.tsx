import { useEffect, useState } from "react";

import api from "../services/api";

interface UseFetchTvSeriesProps {
  id: string;
}

interface FetchTvSeries {
  data: TvShowDetails | null;
  loading: boolean;
  error: string | null;
}

function useFetchTvSeriesDetails({ id }: UseFetchTvSeriesProps): FetchTvSeries {
  const [data, setData] = useState<TvShowDetails | null>(null);
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    if (!id) return;

    const fetchData = async () => {
      setLoading(true);
      setError(null);

      const url = `shows/${id}`;
      try {
        const response = await api.get<TvShowDetails>(url);
        setData(response.data);
      } catch (err) {
        // setError(err.message);
        setError("en error has ocurred");
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [id]);

  return { data, loading, error };
}

export default useFetchTvSeriesDetails;
