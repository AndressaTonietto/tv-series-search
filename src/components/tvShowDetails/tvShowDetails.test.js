import "@testing-library/jest-dom";

import { render, screen, waitFor } from "@testing-library/react";

import { BrowserRouter as Router } from "react-router-dom";
import TvShowDetails from "./tvShowDetails";
import renderer from "react-test-renderer";
import useFetchTvSeriesDetails from "../../hooks/useFetchTvSeriesDetails";

jest.mock("../../hooks/useFetchTvSeriesDetails");

const testShow = {
  id: 1,
  name: "The IT Crowd",
  rating: { average: 8.7 },
  image: { original: "test-image-url" },
  genres: ["Drama", "Thriller"],
  language: "English",
  summary: "<p>Test summary</p>",
};
describe("TvShowDetails", () => {
  beforeEach(() => {
    useFetchTvSeriesDetails.mockReturnValue({
      data: testShow,
      loading: false,
      error: null,
    });
  });

  it("renders correctly", async () => {
    const tree = renderer
      .create(
        <Router>
          <TvShowDetails />
        </Router>
      )
      .toJSON();
    await waitFor(() => expect(tree).toMatchSnapshot());
  });

  it("renders the show details correctly", async () => {
    render(
      <Router>
        <TvShowDetails />
      </Router>
    );

    await waitFor(() => {
      expect(screen.getByText(testShow.name)).toBeInTheDocument();
      expect(
        screen.getByText(testShow.rating.average.toString())
      ).toBeInTheDocument();
      expect(screen.getByAltText(`${testShow.name} poster`)).toHaveAttribute(
        "src",
        testShow.image.original
      );
      expect(screen.getByText("Test summary")).toBeInTheDocument();
      expect(screen.getByText("Language:")).toBeInTheDocument();
      expect(screen.getByText(testShow.language)).toBeInTheDocument();
      testShow.genres.forEach((genre) => {
        expect(screen.getByText(genre)).toBeInTheDocument();
      });
      expect(screen.getByText("Go back")).toBeInTheDocument();
    });
  });
});
