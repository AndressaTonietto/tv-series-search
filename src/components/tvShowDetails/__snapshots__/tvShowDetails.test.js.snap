// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TvShowDetails renders correctly 1`] = `
[
  <button
    className="btn btn-neutral btn-sm mt-2"
    onClick={[Function]}
  >
    Go back
  </button>,
  <div
    className="divider"
  />,
  <div
    className="md:max-w-2xl lg:max-w-4xl mx-auto"
  >
    <div
      className="flex flex-col mt-4 gap-4 md:mx-10"
    >
      <label
        className="text-xl"
      >
        The IT Crowd
      </label>
      <div
        className="flex items-center gap-2"
      >
        <label>
          Rating:
        </label>
        <div
          className="badge badge-success"
        >
          8.7
        </div>
      </div>
      <div
        className="flex flex-col gap-4 md:block"
      >
        <img
          alt="The IT Crowd poster"
          className="max-w-52 min-h-72 shrink-0 md:float-right"
          src="test-image-url"
        />
        <div
          dangerouslySetInnerHTML={
            {
              "__html": "<p>Test summary</p>",
            }
          }
        />
        <div
          className="flex items-center gap-2 md:mt-4"
        >
          <div
            className="badge badge-accent"
          >
            English
          </div>
          <div
            className="badge badge-secondary"
          >
            Drama
          </div>
          <div
            className="badge badge-secondary"
          >
            Thriller
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
